src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.StringSegment.cs

```diff
@@ -363,6 +360,126 @@ private void WriteStringSegmentData(ReadOnlySpan<byte> escapedValue)
+            }
+            else
+            {
+                WriteBase64StringSegmentData(value, isFinalSegment);
+            }
+
+            if (isFinalSegment)
+            {
+                WriteStringSegmentEpilogue();
+
+                SetFlagToAddListSeparatorBeforeNextItem();
+                PreviousSegmentEncoding = SegmentEncoding.None;
+                _tokenType = JsonTokenType.String;
+            }
+        }
+
+        private void WriteBase64StringSegmentWithLeftover(scoped ReadOnlySpan<byte> bytes, bool isFinalSegment)
+        {
+            Debug.Assert(HasPartialCodePoint);
+            Debug.Assert(PreviousSegmentEncoding == SegmentEncoding.Base64);
+
+            scoped ReadOnlySpan<byte> partialCodePointBuffer = PartialBase64CodePoint;
+
+            Span<byte> combinedBuffer = stackalloc byte[3];
+            combinedBuffer = combinedBuffer.Slice(0, ConcatInto(partialCodePointBuffer, bytes, combinedBuffer));
+            if (combinedBuffer.Length is 3)
+            {
+                // Divide up the code point bytes into its own buffer and the remainder of the input buffer.
+                bytes = bytes.Slice(3 - partialCodePointBuffer.Length);
+                partialCodePointBuffer = combinedBuffer.Slice(0, 3);
+            }
+            else
+            {
+                Debug.Assert(combinedBuffer.Length is 1 or 2);
+                // Need more data. If this is a final segment, then the encoder will append '=' as needed.
+                Debug.Assert(bytes.Length + partialCodePointBuffer.Length < 3);
+                bytes = combinedBuffer;
+                partialCodePointBuffer = [];
+            }
+
+            // It doesn't matter if we pass true or false for isFinalSegment since we are guaranteed to not have a partial code point
+            // here (it is either empty or completed using the combined buffer above).
+            WriteBase64StringSegmentData(partialCodePointBuffer, false);
+
+            ClearPartialCodePoint();
+
+            WriteBase64StringSegmentData(bytes, isFinalSegment);
+        }
+
+        private void WriteBase64StringSegmentData(ReadOnlySpan<byte> bytes, bool isFinal)
+        {
+            int leftoverSize;
+            if (!isFinal && (leftoverSize = bytes.Length % 3) != 0)
+            {
+                // If this is not the final segment, we need to wait for more data to come in.
+                PartialBase64CodePoint = bytes.Slice(bytes.Length - leftoverSize);
+                bytes = bytes.Slice(0, bytes.Length - leftoverSize);
+            }
+
+            if (bytes.Length == 0)
+            {
+                return;
+            }
+
+            int requiredBytes = Base64.GetMaxEncodedToUtf8Length(bytes.Length);
+
+            if (_memory.Length - BytesPending < requiredBytes)
+            {
+                Grow(requiredBytes);
+            }
+
+            Span<byte> output = _memory.Span;
+
+            Base64EncodeAndWrite(bytes, output);
```

Comment:
Given that `isFinal` isn't passed to the encoder, how can it tell if padding characters should be appended?

Reply:
For non-final segments, the input is sliced to be a multiple of 3 bytes (earlier in this method) which guarantees that the base64 encoding will never end with padding since `3x` input bytes turn into exactly `4x` base64 bytes. For final segments, the input is passed in as is and will be padded as needed.

Reply:
Would it be possible to add a comment recording this?